Issue/13328 13329 backup restore UI cycle 2

zwarm opened the pull request:
Parents #13328, #13329 and copy improvements #13916

This PR is part two of UI updates and includes
- Updated copy for Restore and Backup Download
- Updated view for status error
- Updated icon/copy for generic error views
- Split out building error views based on type (RemoteRequestFailure vs. GenericFailure)
- Introduced a new common item type `BACKUP_RESTORE_BULLET` and associated supporting classes `BulletState`, and `JetpackBackupRestoreBulletViewHolder`. Updated builders and adapter
- Added new icons for error views
- Caught and fixed an issue with setting error index for restore in `RestoreStep`

Merge Instructions
- Make sure #13898 has been merged
- Remove Not Ready for Merge Label
- Merge as normal

|restore status error|generic error|
|---|---|
|![status](https://user-images.githubusercontent.com/506707/106365515-2b281c00-6304-11eb-879d-5939cb93bbef.png)|![generic](https://user-images.githubusercontent.com/506707/106365523-31b69380-6304-11eb-9585-d7de6105de62.png)

|backup download status error|generic error|
|---|---|
|![status](https://user-images.githubusercontent.com/506707/106365525-354a1a80-6304-11eb-92ac-cec3350ff3ee.png)|![generic](https://user-images.githubusercontent.com/506707/106365532-3c712880-6304-11eb-8417-dcc3f5577bd1.png)

cc: @osullivanchris 
There weren't any details for backup download status (in #13916), so I winged it - same for the icon colors in status and generic error. Please provide any updates. Thanks. :)

To test:
- Forcing the remote request failure and generic failure states has to be done within the code manually
---------------------------------------------------
- In BackupDownloadViewModel, insert the following as the first two lines `fun buildDetails()`
- `transitionToError(BackupDownloadErrorTypes.RemoteRequestFailure)
        return`
- Build and run the app
- Turn on backup and restore feature flags
- Navigate to Activity Log
- Tap the more menu and tap the Download Backup option
- Note that the Status error view is shown
- Go back and change the two lines to: 
       `transitionToError(BackupDownloadErrorTypes.GenericFailure)
         return`
- Rebuild the app
- Navigate to Activity Log
- Tap the more menu and tap the Download Backup option
- Note that the Generic error view is shown
---------------------------------------------------
- In RestoreViewModel, insert the following as the first two lines `fun buildDetails()`
- `transitionToError(RestoreErrorTypes.RemoteRequestFailure)
        return`
- Build and run the app
- Turn on backup and restore feature flags
- Navigate to Activity Log
- Tap the more menu and tap the Restore to this point option
- Note that the Status error view is shown
- Go back and change the two lines to: 
       `transitionToError(RestoreErrorTypes.GenericFailure)
         return`
- Rebuild the app
- Navigate to Activity Log
- Tap the more menu and tap the Download Backup option
- Note that the Generic error view is shown




PR submission checklist:

- [X] I have considered adding unit tests where possible.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

zwarm committed:
Update strings for restore and backup

zwarm committed:
Add icons for error views

zwarm committed:
Add and update jetpack backup restore copy

zwarm committed:
Add bullet item list type for status error view

zwarm committed:
Add bullet state

zwarm committed:
Split error view building and incorporate bullet state

zwarm committed:
Fix indexForErrorTransition value

zwarm committed:
Update builder fun name

zwarm committed:
Update test

zwarm committed:
Remove resource string

malinajirka commented on WordPress/src/main/res/drawable/ic_cloud_off_white_24dp.xml:
```diff
@@ -0,0 +1,9 @@
+    android:viewportHeight="24">
+  <path
+      android:pathData="M19.35,10.04C18.67,6.59 15.64,4 12,4c-1.48,0 -2.85,0.43 -4.01,1.17l1.46,1.46C10.21,6.23 11.08,6 12,6c3.04,0 5.5,2.46 5.5,5.5v0.5H19c1.66,0 3,1.34 3,3 0,1.13 -0.64,2.11 -1.56,2.62l1.45,1.45C23.16,18.16 24,16.68 24,15c0,-2.64 -2.05,-4.78 -4.65,-4.96zM3,5.27l2.75,2.74C2.56,8.15 0,10.77 0,14c0,3.31 2.69,6 6,6h11.73l2,2L21,20.73 4.27,4 3,5.27zM7.73,10l8,8H6c-2.21,0 -4,-1.79 -4, ...
+      android:fillColor="#ffffff"/>
```
🔍 Nitpick: We might want to consider replacing this with a color reference. Same applies to the other files.

zwarm replied:
You are 100% correct ... in commit 5f1e851

malinajirka replied:
👍 Thanks!

Just in case you missed, there are still two files ic_gridicons_checkmark_circle.xml and ic_query_builder_white_24dp.xml which have hardcoded values. We can update them later.

zwarm replied:
I updated in commit ad74918. Simple enough to do it now.

malinajirka reviewed (approved):
Thanks @zwarm! I've tested the flows using Charles Proxy (I haven't made any code changes). The screens look great ;). 

There is just one thing which confuses me a bit. It's a minor thing, I'm just curious why it behaves this way.
1. Open Activity Log
2. Click on "Download backup" action
3. Confirm the action
4. Turn on airplane mode
5. Notice Generic error is shown - I'd expect to see the RemoteRequestFailure, since we didn't receive any response from the server. Wdyt?

zwarm committed:
Replace hardcoded color with color ref

zwarm committed:
Show status error when network is unavailable while request in progress

zwarm committed:
Show status error when network is unavailable while request in progress

zwarm commented:
> There is just one thing which confuses me a bit. It's a minor thing, I'm just curious why it behaves this way.

🤔   Aah, the network error turns is handled as a Generic Error when it gets past the details view. It makes more sense to always show the Status error view when we are past the details view. I changed this in 03c513b. I'll changed restore to behave the same way in b79e73e

zwarm committed:
Resolve conflict on develop merge

zwarm committed:
Replace hardcoded colors with color reference

osullivanchris commented:
> icon colors

Have you used ones from [this link you gave me in the other PR](https://github.com/wordpress-mobile/WordPress-Android/blob/07d47177ed07618c452ed5e33e2fc035710b84cc/WordPress/src/main/res/values/colors.xml#L98)? I actually didn't know where we would have colors defined. But the success, error, warning colours there should be what we need.

**Type sizes**

Similar to an earlier PR, they look a little small. I've used h6 20dp and body 16dp again. 

**Icons**

I've just bumped them up to 32dp, I think it works better with the keyline. They looked too far from the text otherwise when I used the keyline.

**Layout**

All other layout details below. Note the 2dp in pink. I aligned the icon vertically with the first line of text. But I found that the text was creeping over the top of the icon visually a little bit. So I pushed the text down by 2dp to account for it. It might not play out exactly the same in code as it did in Sketch though.

![specs](https://user-images.githubusercontent.com/50576199/106593527-d4f7ea80-6548-11eb-8b5b-058323c563ae.png)

zwarm committed:
Resolve merge conflict with develop

zwarm committed:
Remove period at end of sentence

zwarm committed:
Remove period at end of sentence

zwarm commented:
> icon colors

   I was already using error and success, I updated the yellow to warning


> **Type sizes**
> Similar to an earlier PR, they look a little small. I've used h6 20dp and body 16dp again.

I did use H6 for the header and textAppearanceBody1 (16dp) for the text, so those are all good. 

> **Icons**
> I've just bumped them up to 32dp, I think it works better with the keyline. They looked too far from the text otherwise when I used the keyline.

I updated them to 32dp. 

> **Layout**
> All other layout details below. Note the 2dp in pink. I aligned the icon vertically with the first line of text. But I found that the text was creeping over the top of the icon visually a little bit. So I pushed the text down by 2dp to account for it. It might not play out exactly the same in code as it did in Sketch though.

With the icon adjustment, it looks much better without the 2 dp down. The sentence will wrap as the screen gets smaller.

Thanks.

zwarm committed:
Adjust styles for error views and add dimens

zwarm committed:
Allow for margins in bullet and subheader

zwarm committed:
Allow for margins in bullet and subheader views

zwarm committed:
Fix lint

zwarm committed:
Resize share link button icon
